### PR TITLE
added trigger executed in migration script to db and added method sup…

### PR DESCRIPTION
**Pull request by johnnieskywalker**
https://app.shortcut.com/oazo-apps/story/3494/handling-triggerexecuted-in-oasis-cache

added trigger executed in migration script to db and added method support in transformer

### CODE REVIEW

File: src/borrow/transformers/automationBotTransformer.ts [hunk 1]
```diff
@@ -62,13 +62,40 @@ async function handleTriggerRemoved(
   );
 }
 
+async function handleTriggerExecuted(
+  params:Dictionary<any>,
+  log: PersistedLog,
+  services: LocalServices,
+  ) {
+    const values = {
```
Comment (johnnieskywalker) [inline src/borrow/transformers/automationBotTransformer.ts:70]: it's the same as in trigger removed so maybe could introduce a type here but I don't see any value added of doing so

File: src/borrow/migrations/019-trigger-executed.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE automation_bot.trigger_executed_events (
+    id  serial primary key,
+    cdp_id                  decimal(78,0) not null,
+    trigger_id decimal(78,0) not null,
+
+    log_index               integer not null,
```
Comment (rkrasiuk) [inline src/borrow/migrations/019-trigger-executed.sql:6]: indent properly pls

File: src/borrow/transformers/automationBotTransformer.ts (hunk 1 above)
Comment (rkrasiuk) [reply src/borrow/transformers/automationBotTransformer.ts:70]: you're right, no need to

File: src/borrow/transformers/automationBotTransformer.ts [hunk 3]
```diff
@@ -60,6 +64,38 @@ async function handleTriggerRemoved(
...
+      trigger_id: params.triggerId.toString(),
+      cdp_id: params.cdpId.toString(),
+  
+      log_index: log.log_index,
+      tx_id: log.tx_id,
+      block_id: log.block_id,
+    };
+
+  await services.tx.none(
+    `INSERT INTO automation_bot.trigger_removed_events(
```
Comment (johnnieskywalker) [inline src/borrow/transformers/automationBotTransformer.ts:88]: :O

File: src/addresses/goerli.json [hunk 4]
```diff
@@ -222,5 +222,5 @@
   "MIP21_LIQUIDATION_ORACLE": "0x362dfE51E4f91a8257B8276435792095EE5d85C3",
   "PROXY_PAUSE_ACTIONS": "0x8D1187FCa9A104211bd25c689C08718AD8730C83",
   "PROXY_DEPLOYER": "0xc9476Fd378de5b0de5C4280D4323f6F89f723c15",
-  "AUTOMATION_BOT": "0x9e0B62A297Cc52E226b45Ed18FCaF82d7A41b511"
+  "AUTOMATION_BOT": "0x8A08e91326Ac93fBcfcFb13912cEDdcb7Fc28f71"
```
Comment (rkrasiuk) [inline src/addresses/goerli.json:225]: how the fuck is this change in a diff?

File: src/borrow/migrations/019-trigger-executed.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        id serial PRIMARY key,
+        trigger_id DECIMAL(78, 0) NOT NULL,
+        vault_closed_event INTEGER NOT NULL REFERENCES vault.multiply_events(id),
+        log_index INTEGER NOT NULL,
+        tx_id INTEGER NOT NULL REFERENCES vulcan2x.transaction(id)
+        ON DELETE CASCADE,
+        block_id INTEGER NOT NULL REFERENCES vulcan2x.block(id)
+        ON DELETE CASCADE,
+        UNIQUE (tx_id, log_index)
+    )
```
Comment (rkrasiuk) [inline src/borrow/migrations/019-trigger-executed.sql:12]: pls format this migration in the same fashion others are

File: src/borrow/transformers/automationBotTransformer.ts [hunk 6]
```diff
@@ -62,25 +62,61 @@ async function handleTriggerRemoved(
...
+  log: PersistedLog,
+  services: LocalServices,
+) {
+  const values = {
+    trigger_id: params.triggerId.toString(),
+
+    log_index: log.log_index,
+    tx_id: log.tx_id,
+    block_id: log.block_id,
+  };
```
Comment (rkrasiuk) [inline src/borrow/transformers/automationBotTransformer.ts:76]: pls, store the execution data as well

File: src/borrow/transformers/automationBotTransformer.ts [hunk 7]
```diff
@@ -62,25 +62,61 @@ async function handleTriggerRemoved(
...
+  const closeEventId = matchingVaultClosedEvent.id;
+
+  await services.tx.none(
+    `INSERT INTO automation_bot.trigger_executed_events(
+        trigger_id, vault_closed_event, log_index, tx_id, block_id
+    ) VALUES (
+        \${trigger_id}, \${closeEventId}, \${log_index}, \${tx_id}, \${block_id}
+    );`,
+    values,
+  );
```
Comment (rkrasiuk) [inline src/borrow/transformers/automationBotTransformer.ts:92]: closeEventId is not set in `values` object, that's this query will always fail

File: src/config.goerli.ts [hunk 8]
```diff
@@ -128,6 +131,13 @@ const automationBot = {
   startingBlock: GOERLI_STARTING_BLOCKS.AUTOMATION_BOT,
 };
 
+const multiply = [
+  {
+    address: '0x24E54706B100e2061Ed67fAe6894791ec421B421',
```
Comment (rkrasiuk) [inline src/config.goerli.ts:136]: please, add it either to `addresses/goerli.json` file or the `addresses` variable inside this file

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: src/addresses/goerli.json (hunk 4 above)
Comment (rkrasiuk) [reply src/addresses/goerli.json:225]: i wonder if github can't diff anymore heh

File: src/borrow/migrations/019-trigger-executed.sql (hunk 5 above)
Comment (johnnieskywalker) [reply src/borrow/migrations/019-trigger-executed.sql]: done maually lmk if there's some formatter for vs code that does it default one is not satisfactory

File: src/borrow/transformers/automationBotTransformer.ts (hunk 6 above)
Comment (johnnieskywalker) [reply src/borrow/transformers/automationBotTransformer.ts]: solved

**Comment by johnnieskywalker** [comment]
As borrow vaults don't have close event this is how it looks in cache 
image

This is event 42 from this tx 
https://goerli.etherscan.io/tx/0x0e04a004827a7454eda4ee78966da94323fe11717249a4d505c6d855c7907419#eventlog

File: src/addresses/goerli.json [hunk 9]
```diff
@@ -222,5 +222,6 @@
   "MIP21_LIQUIDATION_ORACLE": "0x362dfE51E4f91a8257B8276435792095EE5d85C3",
   "PROXY_PAUSE_ACTIONS": "0x8D1187FCa9A104211bd25c689C08718AD8730C83",
   "PROXY_DEPLOYER": "0xc9476Fd378de5b0de5C4280D4323f6F89f723c15",
-  "AUTOMATION_BOT": "0x8A08e91326Ac93fBcfcFb13912cEDdcb7Fc28f71"
+  "AUTOMATION_BOT": "0x8A08e91326Ac93fBcfcFb13912cEDdcb7Fc28f71",
```
Comment (johnnieskywalker) [inline src/addresses/goerli.json:225]: Address will change once new smart contract will be redeployed containing this changes and others 
https://github.com/OasisDEX/automation-smartcontracts/pull/32/files

File: abis/automation-bot.json [hunk 10]
```diff
@@ -88,6 +88,12 @@
            "name":"triggerId",
            "type":"uint256"
         },
+        {
+         "indexed":true,
+         "internalType":"uint256",
+         "name":"cdpId",
+         "type":"uint256"
+        },
```
Comment (johnnieskywalker) [inline abis/automation-bot.json:96]: This ABI was edited manually this is change that will need to be deployed

File: src/borrow/transformers/automationBotTransformer.ts [hunk 11]
```diff
@@ -62,25 +62,63 @@ async function handleTriggerRemoved(
...
+  multiplyProxyActionsAddress: SimpleProcessorDefinition[],
+  
+) => BlockTransformer = (address, multiplyProxyActionsAddress) => {
   const deps = normalizeAddressDefinition(address);
-
+  
   return {
     name: getAutomationBotTransformerName(deps.address),
     dependencies: [getExtractorName(deps.address)],
+    transformerDependencies: [...multiplyProxyActionsAddress.map(mpa => getMultiplyTransformerName(mpa))],
```
Comment (rkrasiuk) [inline src/borrow/transformers/automationBotTransformer.ts:121]: unnecessary spread
```suggestion
    transformerDependencies: multiplyProxyActionsAddress.map(mpa => getMultiplyTransformerName(mpa)),
```

File: src/borrow/transformers/automationBotTransformer.ts [hunk 12]
```diff
@@ -62,25 +62,63 @@ async function handleTriggerRemoved(
...
+  },
 };
 
 export const getAutomationBotTransformerName = (address: string) =>
   `automationBotTransformer-${address}`;
 export const automationBotTransformer: (
   address: string | SimpleProcessorDefinition,
-) => BlockTransformer = address => {
+  multiplyProxyActionsAddress: SimpleProcessorDefinition[],
+
```
Comment (rkrasiuk) [inline src/borrow/transformers/automationBotTransformer.ts:114]: nit: rm this empty line

File: src/borrow/transformers/automationBotTransformer.ts [hunk 13]
```diff
@@ -62,25 +62,63 @@ async function handleTriggerRemoved(
...
+  log: PersistedLog,
+  services: LocalServices,
+) {
+    
+  // Closed event exist for multiply vaults but not for borrow vaults.
+  // User can payback DAI and withdraw collateral, but it doesn't mean vault is closed.
+  // One can deposit collateral and generate again on existing vault. ~ŁW
+  const matchingVaultClosedEvent = await services.tx.oneOrNone(
+    `SELECT * FROM vault.multiply_events me WHERE
+       kind = 'exit_collateral' or kind = 'exit_dai' and tx_id = ${log.tx_id}
```
Comment (rkrasiuk) [inline src/borrow/transformers/automationBotTransformer.ts:76]: fix logical precedence with parenthesis
https://www.postgresql.org/docs/7.2/sql-precedence.html

**Review by rkrasiuk — CHANGES_REQUESTED** [review]

File: src/borrow/transformers/automationBotTransformer.ts (hunk 13 above)
Comment (johnnieskywalker) [reply src/borrow/transformers/automationBotTransformer.ts:76]: thanks for this observation so I guess fix will look like:
```sql
SELECT * FROM vault.multiply_events me WHERE
       (kind = 'exit_collateral' or kind = 'exit_dai') and tx_id = ${log.tx_id}
      LIMIT 1;
```

**Review by rkrasiuk — APPROVED** [review]

**Pull request merged**